ospfd/ospf_gr_helper.c

```diff
@@ -0,0 +1,982 @@
+			 enum ospf_helper_exit_reason reason)
+{
+	struct ospf_interface *oi = nbr->oi;
+	struct ospf *ospf = oi->ospf;
+
+	if (!I_AM_HELPER(nbr))
+		return;
+
+	if (IS_DEBUG_OSPF_GR_HELPER)
+		zlog_debug("%s, Exiting from HELPER support to %s, due to %s",
+			   __PRETTY_FUNCTION__, inet_ntoa(nbr->src),
+			   ospf_exit_reason_desc[reason]);
+
+	/* Reset helper status*/
+	nbr->gr_helper_info.gr_helper_status = OSPF_GR_NOT_HELPER;
+	nbr->gr_helper_info.helper_exit_reason = reason;
+	nbr->gr_helper_info.actual_grace_period = 0;
+	nbr->gr_helper_info.recvd_grace_period = 0;
+	nbr->gr_helper_info.gr_restart_reason = 0;
+	ospf->last_exit_reason = reason;
+
+	if (ospf->active_restarter_cnt <= 0) {
+		zlog_err(
+			"OSPF GR-Helper: active_restarter_cnt should be greater than zero here.");
+		return;
+	}
+	/* Decrement active Restarter count */
+	ospf->active_restarter_cnt--;
+
+	/* If the exit not triggered due to grace timer
+	 * expairy , stop the grace timer.
+	 */
+	if (reason != OSPF_GR_HELPER_GRACE_TIMEOUT)
+		THREAD_OFF(nbr->gr_helper_info.t_grace_timer);
+
+	/* check exit triggered due to successfull completion
+	 * of gracefull restart.
+	 * If no, bringdown the neighbour.
+	 */
+	if (reason != OSPF_GR_HELPER_COMPLETED) {
+		if (IS_DEBUG_OSPF_GR_HELPER)
+			zlog_debug(
+				"%s, Unsuccessfull GR exit, so bringing down the neighbour",
+				__PRETTY_FUNCTION__);
+		OSPF_NSM_EVENT_EXECUTE(nbr, NSM_KillNbr);
+	}
+
+	/*Recalculate the DR for the network segment */
+	ospf_dr_election(oi);
+
+	/* Originate a router LSA */
+	ospf_router_lsa_update_area(oi->area);
+
+	/* Originate network lsa if it is an DR in the LAN */
+	if (oi->state == ISM_DR)
+		ospf_network_lsa_update(oi);
+}
+
+/*
+ * Process Maxage Grace LSA.
+ * It is a indication for successfull completion of GR.
+ * If router acting as HELPER, It exits from helper role.
+ *
+ * ospf
+ *    Ospf pointer.
+ *
+ * lsa
+ *    Grace LSA received from RESTARTER.
+ *
+ * nbr
+ *    ospf neighbour which requets the router to act as
+ *    HELPER.
+ *
+ * Returns:
+ *    Nothing.
+ */
+void ospf_process_maxage_grace_lsa(struct ospf *ospf, struct ospf_lsa *lsa,
+				   struct ospf_neighbor *nbr)
+{
+	struct in_addr restartAddr = {0};
+	uint8_t restartReason = 0;
+	uint32_t graceInterval = 0;
+	struct ospf_neighbor *restarter = NULL;
+	struct ospf_interface *oi = nbr->oi;
+	int ret;
+
+	/* Extract the grace lsa packet fields */
+	ret = ospf_extract_grace_lsa_fields(lsa, &graceInterval, &restartAddr,
+					    &restartReason);
+	if (ret != OSPF_SUCCESS) {
+		if (IS_DEBUG_OSPF_GR_HELPER)
+			zlog_debug("%s, Wrong Grace LSA packet.",
+				   __PRETTY_FUNCTION__);
+		return;
+	}
+
+	if (IS_DEBUG_OSPF_GR_HELPER)
+		zlog_debug("%s, GraceLSA received for neighbour %s.",
+			   __PRETTY_FUNCTION__, inet_ntoa(restartAddr));
+
+	/* Incase of broadcast links, if RESTARTER is DR_OTHER,
+	 * grace LSA might be received from DR, so need to get
+	 * actula neighbour info , here RESTARTER.
```

Comment:
actual

I don't understand the last part about RESTARTER, can you elaborate?

Reply:
If the RESTARTER is non-DR, HELPER receives graces lsa packets from DR in the network so here nbr pointer should be populated based on RESTARTER address from the packet.